### DISCUSSION TITLE
Different approach to NVMe subsystem handling

### DISCUSSION THREAD

**Discussion by matteotenca** [pkoutoupis/rapiddisk#163] [Ideas]
Hi @pkoutoupis !

I was thinking of a new set of switch to handle the creation/deletion of ports, allowed hosts and NQN, and to link them to each other independently.

1. A switch to create/remove a port, which require to specify the ip OR the network interface​ AND the port number AND the protocol on creation, just the port on deletion (since a port can be mapped to an IP only, if I am right).
Examples (switch names are casual):

   - `$ rapiddisk -i -P port_number -i interface|ip_address -t tcp|rdma|loop`
   - `$ rapiddisk -X -P port_number`

2) A switch to add/remove allowed hostname/s globally
   - `$ rapiddisk -y -H hostname[,hostname,..]`
   - `$ rapiddisk -Y -H hostname[,hostname,..]`

3) A switch to add/remove a NQN without linking it to any port/allowed host/s
   - `$ rapiddisk -e -b ramdisk_name`
   - `$ rapiddisk -x -b ramdisk_name`
   
4) A switch to link/unlink an existing port to an existing NQN:
   - `$ rapiddisk -z -P port[,port,..] -b ramdisk_name`
   - `$ rapiddisk -Z -P port[,port,..] -b ramdisk_name`
   
5) A switch to link/unlink an existing allowed host/s to an existing NQN
   - `$ rapiddisk -w -H hostname[,hostname,..] -b ramdisk_name`
   - `$ rapiddisk -W -H hostname[,hostname,..] -b ramdisk_name`

We can decide what to do when the user tries to remove an NQN/port/hosts with linked NQN/ports/hosts (for example we can add a force switch which will unlink a NQN from all the used resources and then remove the NQN itself, or just return some information about the current configuration).

Tell me your thoughts

​Cheers!!
Teo

**Comment by pkoutoupis** [comment]
For tracking purposes: https://github.com/pkoutoupis/rapiddisk/issues/148

Thank you for continuing this dialogue. I will review shortly.

**Comment by pkoutoupis** [comment]
Thinking this through: why couldn't numbers 2 and 5 be together? For example:

- Number 2 would return with "host NQN in use", if it is still connected as an allowed host to any target NQN.
- Number 5 could, in theory (which you already built the functionality), just automatically remove any unused host NQN.

**Reply by matteotenca** [reply]
> Thinking this through: why couldn't numbers 2 and 5 be together? For example:
> 
>     * Number 2 would return with "host NQN in use", if it is still connected as an allowed host to any target NQN.
> 
>     * Number 5 could, in theory (which you already built the functionality), just automatically remove any unused host NQN.

Because by doing so we would fall back into the old "mixed" approach. Keeping "2" and "5" separate creates a hierarchy between commands: some can only be executed if another has already been executed.

As an extreme example, we could combine `-a` and `-m`, or `-d` and `-u`, for example: `rapiddisk -d rd0` could automatically remove mappings using `rd0`, all NVMe exports using `rd0 `, and any hosts required by those exports, and ports as well. Needless to say, it's a nightmare to code this approach cleanly, for all the cross-checking that would have to be done to avoid breaking other mappings or exports - even if it would somehow make perfect sense.

By keeping all actions separate, we can write smaller functions that perform atomic tasks, and from a user perspective the chain of actions is clear: NQNs are something, hosts are something else, ports are something else, and , once created, these three entities can be related using simple atomic operations.

If you recall, during the call we both spent some time staring at the code imagining the various scenarios that could have occurred during the NVMe unexport - and in the end luckily you noticed that yet another check was missing (to avoid deleting a port if used by another NQN). I'd like to divide all the command line activities so as not to have to write convoluted and complex functions forced to consider four/five different scenarios in order not to run into huge errors.

Let me know what you think!

**Reply by pkoutoupis** [reply]
@matteotenca OK. You convinced me. Who did you want to do the work? I have no issue doing the implementation but knowing you, it is either already implemented in a branch or part of the way there. :-)

**Comment by matteotenca** [comment]
> @matteotenca OK. You convinced me. Who did you want to do the work? I have no issue doing the implementation but knowing you, it is either already implemented in a branch or part of the way there. :-)

@pkoutoupis Eheheheheheheheehhe I can do it, if that's okay with you.!

This time I did not went straight with the implementation, but I spent some time to realize the autocompletion script for the current `rapiddisk` tool... You can find it attached to this message, to test it just `source` the file from a shell:

```bash
shub@ubuserver:~$ source ./rapiddisk_autocomplete.sh
```

If you have time, tell me what you think! :-)

**Reply by pkoutoupis** [reply]
Of course it is okay with me. I just do not want to put any pressure on you.

Re: autocompletion tool....
Neat! I will look at the script and try it out shortly!

**Comment by matteotenca** [comment]
```bash
#/usr/bin/env bash

# VERSION=0.0.1
# Copyright 2023 Matteo Tenca <matteo.tenca@gmail.com>
# License: GPL v2

_rapiddisk_completions()
{
	local cur prev opts first second third fourth opt1 opt2

	new_array=()

	# Iterate over the original array
	for item in "${COMP_WORDS[@]}"; do
		if [[ "$item" != "-j" ]]; then
			# Add the item to the new array if it's not "-j"
			new_array+=("$item")
		fi
	done

	first="${new_array[1]}"
	opt1="${new_array[2]}"
	second="${new_array[3]}"
	opt2="${new_array[4]}"
	third="${new_array[5]}"
	opt3="${new_array[6]}"
	fourth="${new_array[7]}"
	opt4="${new_array[8]}"


#	echo -e "\nfirst: '$first'\nsecond: '$second'\nthird: '$third'\nopt1: '$opt1'\nword being completed: '$2'\nprevious: '$prev'\nCOMP_CWORD: '$COMP_CWORD'\nCOMP_WORDS: '${COMP_WORDS[@]}'\ncur: '$cur'\n1: '$1'\n2: '$2'\n3: '$3'"

	if [ $COMP_CWORD -eq 1 ] ; then 
		opts="-j -x -d -h -l -a -r -m -u -f -L -U -i -X -e -R -n -N"
	else
		case $3 in
			-j)
				if [ "${COMP_WORDS[1]}" = "-j" ] && [ $COMP_CWORD -eq 2 ] ; then
					opts="-x -d -l -a -r -m -u -f -L -U -i -X -e -R -n -N"
				fi
				;;
			-d|-r|-m|-f|-U|-L)
				opts=`find /dev -maxdepth 1 -group disk |grep '/dev/rd' | grep -oP 'rd\d+'`
				;;
			-u)
				opts=`ls /dev/mapper/rc-*| sed 's,/dev/mapper/,,g'`
				;;
			-R)
				opts="-b"
				;;
			-i)
				opts=`ip link|grep -oP '^\d+:\s+[^:]+'|grep -Pv '\slo$'|grep -oP '[^\s]+$'`" NULL"
				;;
			-b)
				if [ "$first" = "-e" ] || [ "$first" = "-x" ] || [ "$first" = "-R" ] ; then
					opts=`find /dev -maxdepth 1 -group disk |grep '/dev/rd' | grep -oP 'rd\d+'`
				else
					opts1=`find /dev -maxdepth 1 -group disk`
                    opts=${opts1}' '`ls /dev/mapper/rc-*`

				fi
				;;
			-p)
				opts="wt wa wb"
				;;
			-t)
				if [ "$opt1" != "NULL" ] ; then
					opts="tcp rdma"
				else
					opts="loop"
				fi
				;;
			-X)
				opts="-P"
				;;
			-e|-x)
				opts="-b -P -H"
				;;
			-c)
				;;
			*)
				if [ -n "$first" ] ; then
					case $first in
						-r)
							opts="-c"
							;;
						-m)
							if [ -z "$third" ] || [ "$third" = "-"  ] ; then
								case $second in
									-b)
										if [ "$2" = "-b" ] ; then opts="-b" ; else opts="-p" ; fi
										;;
									-p)
										if [ "$2" = "-p" ] ; then opts="-p" ; else opts="-b" ; fi
										;;
									*)
										opts="-b -p" 
										;;
								esac
							fi
							;;
						-i)
							if [ -z "$third" ] || [ "$third" = "-"  ] ; then
                                case $second in
                                    -P)
                                        if [ "$2" = "-P" ] ; then opts="-P" ; elif [ -n "$opt2" ] ;  then opts="-t" ; fi
                                        ;;
                                    -t)
                                        if [ "$2" = "-t" ] ; then opts="-t" ; else opts="-P" ; fi
                                        ;;
                                    *)
                                		opts="-P -t"
                                        ;;
                                esac
                            fi
							;;
						-e|-x)
							if [ -z "$opt2" ] || [ "$opt2" = "-" ] ; then
                                case $opt1 in
                                    -P)
                                        if [ "$2" = "-P" ] ; then opts="-P" ; elif [ -n "$second" ] ; then opts="-H -b" ; fi 
                                        ;;
                                    -H)
                                        if [ "$2" = "-H" ] ; then opts="-H" ; elif [ -n "$second" ] ; then opts="-P -b" ; fi
                                        ;;
                                    -b)
                                        if [ "$2" = "-b" ] ; then opts="-b" ; elif [ -n "$second" ] ; then opts="-P -H " ; fi
                                        ;;
                                    *)
                                        ;;
                                esac
                            fi
							if [ -z "$opt3" ] || [ "$opt3" = "-" ] ; then
                                case $opt2 in
                                    -P)
                                        if [ "$2" = "-P" ] ; then opts="-P" ; elif [ -n "$third" ] ; then if [ "$opt1" = "-H" ] ; then opts="-b" ; else opts="-H"; fi ; fi
                                        ;;
                                    -H)
                                        if [ "$2" = "-H" ] ; then opts="-H" ; elif [ -n "$third" ] ; then if [ "$opt1" = "-P" ] ; then opts="-b" ; else opts="-P"; fi ; fi
                                        ;;
                                    -b)
                                        if [ "$2" = "-b" ] ; then opts="-b" ; elif [ -n "$third" ] ; then if [ "$opt1" = "-H" ] ; then opts="-P" ; else opts="-H"; fi ; fi
                                        ;;
                                    *)
                                        ;;
                                esac
                            fi

							;;
						*)
							;;
					esac
				fi
				;;
		esac
	fi
	COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    COMPREPLY=( $(compgen -W "$opts" -- ${cur}) )
    return 0

}

complete -F _rapiddisk_completions rapiddisk

```

**Reply by pkoutoupis** [reply]
@matteotenca I like this. I played with it for a little bit. I look forward to when you open up a PR. Some thoughts on location:
```
git mv scripts/rapiddisk-legacy scripts/misc
git mv scripts/misc/README.md scripts/misc/rapiddisk-legacy.txt
cp rapiddisk_autocomplete.sh scripts/misc/
```
And then create a new `scripts/misc/README.md` to describe the collection of scripts and installation/sourcing. Thoughts?

**Reply by matteotenca** [reply]
I believe it would be fine if the autocompletion file is installed via `Make`... We could put it in the `doc` dir, or a subdir of `doc`.

**Reply by pkoutoupis** [reply]
I do not mind having it installed by `make` (I guess the question is: install it to where?). I still feel that maybe we need to place it in the script subdirectory. I do not know how it qualifies as a document or document-type file to be placed in the `doc` subdirectory. Either way, I do enjoy what I am seeing with this. Thank you again!

**Comment by matteotenca** [comment]
@pkoutoupis Here is a very first implementation:

https://github.com/matteotenca/rapiddisk/tree/new_options

**Reply by matteotenca** [reply]
I was on a rush, here are some more info:

1. I did not creat a PR yet because it is a rough implementation. It keeps all the old switches and functions, I added some new (temp) switches to implement the new approach. If it is easier for you to check out the new code with a PR, I will submit one.
2. The new messages and strings for error and messages probably contain grammar/syntax errors.
3. I tested the NQN/port/host configs created via the new code, and the exports seem to work, I am not completely sure about the presence of all the bells and whistles needed by the NVMe subsystem though.

Let me know what you think! :-)

**Reply by pkoutoupis** [reply]
@matteotenca Thank you for you hard work and persistence in keeping this project both stable and relevant. I will check it out shortly.

**Reply by pkoutoupis** [reply]
@matteotenca Any way you can create that PR? I would like to start commenting.

**Comment by matteotenca** [comment]
> @matteotenca Any way you can create that PR? I would like to start commenting.

@pkoutoupis I was out of town for a month now, I'll be back this weekend and I will manage to get the job done ASAP. Sorry for the long delay!

**Reply by pkoutoupis** [reply]
No worries and no need to be sorry. I hope that you had a nice and relaxing time away. 😄

**Comment by pkoutoupis** [comment]
https://github.com/pkoutoupis/rapiddisk/pull/173
